libraries/cyclestreets-view/src/main/java/net/cyclestreets/views/overlay/TapToRouteOverlay.kt

```diff
@@ -0,0 +1,352 @@
+                else
+                    return false
+        }
+
+        return true // we handled it!
+    }
+
+    ////////////////////////////////////////////
+    override fun draw(canvas: Canvas, mapView: MapView, shadow: Boolean) {
+        drawRoutingInfoRect()
+        drawRestartButton()
+    }
+
+    private fun drawRestartButton() {
+        if (tapState.routeIsPlanned()) {
+            restartButton.show()
+        } else {
+            restartButton.hide()
+        }
+    }
+
+    private fun drawRoutingInfoRect() {
+        if (tapState.routeIsPlanned()) {
+            // In this case, populating the routing info is done by the RouteHighlightOverlay
+            return
+        }
+
+        routeNowIcon.visibility = if (tapState.canRoute()) View.VISIBLE else View.INVISIBLE
+
+        routingInfoRect.apply {
+            setBackgroundColor(if (tapState.canRoute()) highlightColour else lowlightColour)
+            gravity = Gravity.CENTER
+            text = tapState.actionDescription
+            isEnabled = tapState.canRoute()
+        }
+    }
+
+    //////////////////////////////////////////////
+    override fun onSingleTap(event: MotionEvent): Boolean {
+        return tapMarker(event)
+    }
+
+    override fun onDoubleTap(event: MotionEvent): Boolean {
+        return false
+    }
+
+    private fun tapRestart(): Boolean {
+        if (!CycleStreetsPreferences.confirmNewRoute())
+            return stepBack(true)
+
+        MessageBox.YesNo(mapView,
+                "Start a new route?"
+        ) { _, _ -> stepBack(true) }
+
+        return true
+    }
+
+    override fun onBackPressed(): Boolean {
+        return stepBack(false)
+    }
+
+    private fun stepBack(tap: Boolean): Boolean {
+        if (!tap && !tapState.waypointingInProgress)
+            return false
+
+        when (tapState) {
+            TapToRoute.WAITING_FOR_START -> return true
+            TapToRoute.WAITING_TO_ROUTE,
+            TapToRoute.WAITING_FOR_SECOND,
+            TapToRoute.WAITING_FOR_NEXT -> waymarks.removeWaypoint()
+            TapToRoute.ALL_DONE -> Route.resetJourney()
+        }
+
+        tapState = tapState.previous(waypointsCount())
+        mapView.postInvalidate()
+
+        return true
+    }
+
+    private fun tapMarker(event: MotionEvent): Boolean {
+        val p = mapView.projection.fromPixels(event.x.toInt(), event.y.toInt())
+        tapAction(p)
+        return true
+    }
+
+    fun setNextMarker(point: IGeoPoint) {
+        tapAction(point)
+    }
+
+    private fun tapAction(point: IGeoPoint) {
+        if (tapState.noFurtherWaypoints()) {
+            return
+        }
+
+        waymarks.addWaypoint(point)
+        controller.pushUndo(this)
+        tapState = tapState.next(waypointsCount())
+        mapView.invalidate()
+    }
+
+    ////////////////////////////////////
+    private enum class TapToRoute private constructor(val waypointingInProgress: Boolean, val actionDescription: String) {
```

Comment:
I wonder about moving this enum out into its own file in the same package - as an `internal enum` rather than a `private enum`.  What do you think?